streams/src/main/java/org/apache/kafka/streams/state/internals/PrefixedWindowKeySchemas.java

```diff
@@ -0,0 +1,311 @@
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.state.internals;
+
+import org.apache.kafka.common.serialization.Deserializer;
+import org.apache.kafka.common.serialization.Serializer;
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.streams.kstream.Window;
+import org.apache.kafka.streams.kstream.Windowed;
+import org.apache.kafka.streams.kstream.internals.TimeWindow;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.nio.ByteBuffer;
+import java.util.List;
+
+import static org.apache.kafka.streams.state.StateSerdes.TIMESTAMP_SIZE;
+import static org.apache.kafka.streams.state.internals.WindowKeySchema.timeWindowForSize;
+
+public class PrefixedWindowKeySchemas {
+
+    private static final int PREFIX_SIZE = 1;
+    private static final byte TIME_FIRST_PREFIX = 1;
+    private static final byte KEY_FIRST_PREFIX = 2;
+    private static final int SEQNUM_SIZE = 4;
+    private static final int SUFFIX_SIZE = TIMESTAMP_SIZE + SEQNUM_SIZE;
+    private static final byte[] MIN_SUFFIX = new byte[SUFFIX_SIZE];
+
+    private static byte extractPrefix(final byte[] binaryBytes) {
+        return binaryBytes[0];
+    }
+
+    public static class TimeFirstWindowKeySchema implements RocksDBSegmentedBytesStore.KeySchema {
+
+        @Override
+        public Bytes upperRange(final Bytes key, final long to) {
+            if (to == Long.MAX_VALUE) {
+                return null;
+            }
+
+            return Bytes.wrap(ByteBuffer.allocate(PREFIX_SIZE + TIMESTAMP_SIZE)
+                .put(TIME_FIRST_PREFIX)
+                .putLong(to + 1)
+                .array());
+        }
+
+        @Override
+        public Bytes lowerRange(final Bytes key, final long from) {
+            if (key == null) {
+                return Bytes.wrap(ByteBuffer.allocate(PREFIX_SIZE + TIMESTAMP_SIZE)
+                    .put(TIME_FIRST_PREFIX)
+                    .putLong(from)
+                    .array());
+            }
+
+            /*
+             * Larger timestamp or key's byte order can't be smaller than this lower range. Reason:
+             *     1. Timestamp is fixed length (with big endian byte order). Since we put timestamp
+             *        first, larger timestamp will have larger byte order.
+             *     2. If timestamp is the same but key (k1) is larger than this lower range key (k2):
+             *         a. If k2 is not a prefix of k1, then k1 will always have larger byte order no
+             *            matter what seqnum k2 has
+             *         b. If k2 is a prefix of k1, since k2's seqnum is 0, after k1 appends seqnum,
+             *            it will always be larger than (k1 + seqnum).
+             */
+            return Bytes.wrap(ByteBuffer.allocate(PREFIX_SIZE + TIMESTAMP_SIZE + key.get().length + SEQNUM_SIZE)
+                .put(TIME_FIRST_PREFIX)
+                .putLong(from)
+                .put(key.get())
+                .putInt(0)
```

Comment:
> so if [timestamp, lower-range] is a prefix of [timestamp, key] then the former would still be put ahead of the latter.

I'm not total clear about this. Is it not the behavior we want: `lower-range` before `key`? Can you give an example? I will also add tests.

Reply:
Basically I was wondering if it is really necessary to pad four 0 bytes on the key schema.

We use `seqnum` for window stores that are used in stream-stream joins, since we need to maintain values of the same `[key, timestamp]`, but for aggregation windowed key we would not need to maintain uniqueness, and hence not necessary to use `seqnum`. I'm basically asking if just `[from, key]` is good enough as the lower range.

Reply:
I think you are right. We can drop `seqnum` here.